Error when making source code on Ubuntu 16.04

i followed these instructions (for linux x86): https://github.com/Tencent/ncnn/wiki/how-to-build

when i tried to build the NCNN using **sudo make -j4**, i got these errors 

![image](https://user-images.githubusercontent.com/34804280/75039512-cf4df400-54f3-11ea-8fa6-a129977908a2.png)

The errors are very similar to this:
**/ncnn/tools/quantize/ncnn2table.cpp:820:26: error: passing ‘const cv::CommandLineParser’ as ‘this’ argument discards qualifiers [-fpermissive]**

im using g++-5, gcc5 and opencv 2.4

Thanks for the quick reply. There are still errors after running your updated git code. 

Here are the screenshots of the error:
Screenshot 1
![image](https://user-images.githubusercontent.com/34804280/75040573-db3ab580-54f5-11ea-8cbd-319f546745b7.png)
Screenshot 2
![image](https://user-images.githubusercontent.com/34804280/75040600-ebeb2b80-54f5-11ea-97b7-94024522f478.png)

The full output is as follows:
**[ 71%] Linking CXX executable ncnnoptimize
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp: In function ‘int main(int, char**)’:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:820:26: error: within this context
     if (parser.has("help"))
                          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:822:16: error: ‘class cv::CommandLineParser’ has no member named ‘printMessage’
         parser.printMessage();
                ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:827:28: error: within this context
     if (!parser.has("param") || !parser.has("bin") || !parser.has("images") || !parser.has("output") || !parser.has("mean") || !parser.has("norm"))
                            ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:827:50: error: within this context
     if (!parser.has("param") || !parser.has("bin") || !parser.has("images") || !parser.has("output") || !parser.has("mean") || !parser.has("norm"))
                                                  ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:827:75: error: within this context
     if (!parser.has("param") || !parser.has("bin") || !parser.has("images") || !parser.has("output") || !parser.has("mean") || !parser.has("norm"))
                                                                           ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:827:100: error: within this context
     if (!parser.has("param") || !parser.has("bin") || !parser.has("images") || !parser.has("output") || !parser.has("mean") || !parser.has("norm"))
                                                                                                    ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:827:123: error: within this context
     if (!parser.has("param") || !parser.has("bin") || !parser.has("images") || !parser.has("output") || !parser.has("mean") || !parser.has("norm"))
                                                                                                                           ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:827:146: error: within this context
     if (!parser.has("param") || !parser.has("bin") || !parser.has("images") || !parser.has("output") || !parser.has("mean") || !parser.has("norm"))
                                                                                                                                                  ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:830:16: error: ‘class cv::CommandLineParser’ has no member named ‘printMessage’
         parser.printMessage();
                ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:860:26: error: within this context
     if (parser.has("mean"))
                          ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:878:26: error: within this context
     if (parser.has("norm"))
                          ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:896:26: error: within this context
     if (parser.has("size"))
                          ^
In file included from /usr/include/opencv2/opencv.hpp:47:0,
                 from /media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:30:
/usr/include/opencv2/core/core.hpp:4744:10: error: ‘bool cv::CommandLineParser::has(const string&)’ is protected
     bool has(const std::string& keys);
          ^
/media/leejiayi/DATA/ubuntu/Downloads/JobsApplication/BlackSesameTech/ncnn/ncnn/tools/quantize/ncnn2table.cpp:921:28: error: within this context
     if (parser.has("swapRB"))
                            ^
[ 71%] Built target test_softmax
tools/quantize/CMakeFiles/ncnn2table.dir/build.make:62: recipe for target 'tools/quantize/CMakeFiles/**

--- a/src/layer/x86/cast_x86.cpp
+++ b/src/layer/x86/cast_x86.cpp
@@ -22,6 +22,7 @@
 #endif // __AVX__
 
 #if __AVX__
+#include <stdint.h>
 typedef union m128i
 {
     __m128i vec;

--- a/src/layer/x86/cast_x86.cpp
+++ b/src/layer/x86/cast_x86.cpp
@@ -137,7 +137,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
         for (int i = 0; i < remain; i++)
             mask.m256i_u32[i] = 0x80000000;
 
-        #pragma omp parallel for num_threads(opt.num_threads)
+#pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const float* ptr = bottom_blob.channel(q);
@@ -169,7 +169,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
         for (int i = 0; i < remain; i++)
             mask.m256i_u32[i] = 0x80000000;
 
-        #pragma omp parallel for num_threads(opt.num_threads)
+#pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const unsigned short* ptr = bottom_blob.channel(q);
@@ -195,7 +195,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
     }
     if (type_from == 4 && type_to == 1)
     {
-        #pragma omp parallel for num_threads(opt.num_threads)
+#pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const unsigned short* ptr = bottom_blob.channel(q);
@@ -220,7 +220,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
     }
     if (type_from == 1 && type_to == 4)
     {
-        #pragma omp parallel for num_threads(opt.num_threads)
+#pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const float* ptr = bottom_blob.channel(q);

--- a/src/layer/x86/cast_x86.cpp
+++ b/src/layer/x86/cast_x86.cpp
@@ -137,7 +137,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
         for (int i = 0; i < remain; i++)
             mask.m256i_u32[i] = 0x80000000;
 
-#pragma omp parallel for num_threads(opt.num_threads)
+        #pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const float* ptr = bottom_blob.channel(q);
@@ -169,7 +169,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
         for (int i = 0; i < remain; i++)
             mask.m256i_u32[i] = 0x80000000;
 
-#pragma omp parallel for num_threads(opt.num_threads)
+        #pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const unsigned short* ptr = bottom_blob.channel(q);
@@ -195,7 +195,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
     }
     if (type_from == 4 && type_to == 1)
     {
-#pragma omp parallel for num_threads(opt.num_threads)
+        #pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const unsigned short* ptr = bottom_blob.channel(q);
@@ -220,7 +220,7 @@ int Cast_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) 
     }
     if (type_from == 1 && type_to == 4)
     {
-#pragma omp parallel for num_threads(opt.num_threads)
+        #pragma omp parallel for num_threads(opt.num_threads)
         for (int q = 0; q < channels; q++)
         {
             const float* ptr = bottom_blob.channel(q);

--- a/tools/quantize/ncnn2table.cpp
+++ b/tools/quantize/ncnn2table.cpp
@@ -795,6 +795,23 @@ static int find_all_value_in_string(const std::string& values_string, std::vecto
     return 0;
 }
 
+#if CV_MAJOR_VERSION < 3
+class NcnnQuantCommandLineParser: public cv::CommandLineParser
+{
+public:
+    NcnnQuantCommandLineParser(int argc, const char *const argv[], const char *key_map):
+        cv::CommandLineParser(argc, argv, key_map)
+    {
+    }
+    bool has(const std::string &keys) {
+        return cv::CommandLineParser::has(keys);
+    }
+    void printMessage() {
+        cv::CommandLineParser::printParams();
+    }
+};
+#endif
+
 int main(int argc, char** argv)
 {
     std::cout << "--- ncnn post training quantization tool --- " << __TIME__ << " " << __DATE__ << std::endl;
@@ -810,7 +827,11 @@ int main(int argc, char** argv)
                           "{swapRB c       |   | flag which indicates that swap first and last channels in 3-channel image is necessary }"
                           "{thread t       | 4 | count of processing threads }";
 
+#if CV_MAJOR_VERSION < 3
+    NcnnQuantCommandLineParser parser(argc, argv, key_map);
+#else
     cv::CommandLineParser parser(argc, argv, key_map);
+#endif
 
     if (parser.has("help"))
     {

--- a/tools/quantize/ncnn2table.cpp
+++ b/tools/quantize/ncnn2table.cpp
@@ -796,17 +796,19 @@ static int find_all_value_in_string(const std::string& values_string, std::vecto
 }
 
 #if CV_MAJOR_VERSION < 3
-class NcnnQuantCommandLineParser: public cv::CommandLineParser
+class NcnnQuantCommandLineParser : public cv::CommandLineParser
 {
 public:
-    NcnnQuantCommandLineParser(int argc, const char *const argv[], const char *key_map):
-        cv::CommandLineParser(argc, argv, key_map)
+    NcnnQuantCommandLineParser(int argc, const char* const argv[], const char* key_map)
+        : cv::CommandLineParser(argc, argv, key_map)
     {
     }
-    bool has(const std::string &keys) {
+    bool has(const std::string& keys)
+    {
         return cv::CommandLineParser::has(keys);
     }
-    void printMessage() {
+    void printMessage()
+    {
         cv::CommandLineParser::printParams();
     }
 };

